### ISSUE TITLE
Add indicators database objects

### ISSUE THREAD

**Issue by wlthomson** [openmsupply/mobile#1823]
## Is your feature request related to a problem? Please describe.

Add datatypes for 4.08 indicators. 

## Describe the solution you'd like

A direct mapping between desktop and mobile:

- `indicatorValue`
- `indicatorAttribute`
- `programData`

### Indicator

```
ProgramIndicator.schema = {
  name: 'ProgramIndicator',
  primaryKey: 'id',
  properties: {
    id: 'string',
    code: { type: 'string', default: 'placeholderCode' },
    program: 'MasterList',
    isActive: { type: 'bool', default: false }
  }
}
```

### IndicatorAttribute

```
IndicatorAttribute.schema = {
  name: 'IndicatorAttribute',
  primaryKey: 'id',
  properties: {
    id: 'string',
    indicator: 'Indicator',
    description: { type: 'string', default: 'placeholderDescription' },
    code: { type: 'string', default: 'placeholderCode' },
    index: { type: 'int', default: 0 },
    isRequired: { type: 'bool', default: true },
    valueType: { type: 'string', default: 'string' },
    valueDefault: { type: 'string', default: 'string' },
    axis: { type: 'string': default: 'row' },
    is_active: { type: 'bool', default: false }
  }
}
```

### IndicatorValue 

```
IndicatorValue.schema = {
  name: 'IndicatorValue',
  primaryKey: 'id',
  properties: {
    id: 'string',
    facilityId: 'string',
    period: 'Period',
    column: 'IndicatorAttribute'
    row: 'IndicatorAttribute'
    value: type: 'string'
  },
};
```

## Describe alternatives you've considered

N/A.

## Additional context

See epic #1685.

## Discussion:

- Initial brain dump of proposed solution. Feedback/thoughts would be great.
- Much of the info being synced is not editable. Is there an elegant way of making realm models/fields immutable or editable only via a setter? Could be good for some models/fields which should be mobile read-only, e.g. `IndicatorValue.facilityId` (always the mobile store), or `IndicatorAttribute` and `programIndicator` records?

**Comment by joshxg** [comment]
@wlthomson 

1. A decent solution to handle store data storeID, would be to not have the `facilityID` field in realm. Then in outgoingSync, add the field to the request using the stored setting `THIS_STORE_ID`. But this is limited obviously.
2. Might need more defaults etc: Consider the case of receiving an `IndicatorValue` before anything during an initialization. Then you're going to want to create default `IndicatorData` and probably, then, `ProgramIndicator` records. These will need default values. You then also possibly need to have have example, `ProgramIndicator.program` be optional as you don't want to relate this to some random program as you don't actually have the record yet.
3. Might need an optional: value.value?

**Comment by wlthomson** [comment]
>     1. A decent solution to handle store data storeID, would be to not have the `facilityID` field in realm. Then in outgoingSync, add the field to the request using the stored setting `THIS_STORE_ID`. But this is limited obviously.

Yeah, I did think of this, but my thought was it was safer to have a "hands-off" approach and respect the desktop data rather than depend on other logic which could potentially change in the future (i.e. that all values syncing in/out are linked to the mobile store)?

>     2. Might need more defaults etc: Consider the case of receiving an IndicatorValue before anything during an initialization. Then you're going to want to create default IndicatorData and probably, then, ProgramIndicator records. These will need default values. You then also possibly need to have have example, ProgramIndicator.program be optional as you don't want to relate this to some random program as you don't actually have the record yet.

That's a good point. 

>     3. Might need an optional: value.value?

As implemented currently, all `[indicatorValue]` records should be initialised with a value, but agree it's good to handle edge cases. I think it might be best to leave the field as required and handle any "broken" records in incoming sync (the default could be set differently depending on the data type of the parent row/column)?

**Comment by joshxg** [comment]
> it would possibly be better to keep the placeholder pattern consistent and instead of nullifying, create placeholder MasterList records?

The pattern that's been used in the app so far is by having optional links, so I'd argue that just being consistent is best. See: `RequisitionItem.requisition` (it's optional). I think having some static placeholder records for these object types would be great - could have side effects though, as they're not really accounted for. For example, the list of master lists to choose from might have a placeholder master list show up. Just my opinion!

**Comment by wlthomson** [comment]
> > it would possibly be better to keep the placeholder pattern consistent and instead of nullifying, create placeholder MasterList records?
> 
> The pattern that's been used in the app so far is by having optional links, so I'd argue that just being consistent is best. See: `RequisitionItem.requisition` (it's optional). I think having some static placeholder records for these object types would be great - could have side effects though, as they're not really accounted for. For example, the list of master lists to choose from might have a placeholder master list show up. Just my opinion!
> 
> EDIT: Re-reading this it seems like a stupid hypocritical(?)/inconsistent argument. But the pattern of creating placeholders is only applicable when you have access to the ID of the not-yet-receieved record. Without that (i.e. without the master list ID from the value record), you can't make a placeholder)

Sorry, possibly explained poorly. I wasn't suggesting static placeholders, just that instead of making the `MasterList` link optional for an `IndicatorValue`, could just create a placeholder masterlist object in the same way as you suggested for `IndicatorData` etc? All links get synced from desktop with the foreign key, so this should work, right? Of course that doesn't preclude it from being optional, can do both?

**Comment by joshxg** [comment]
Ya sure - I'm just thinking about the value records as the example. `Value` comes in -> have ID for a `Data`, create placeholder. Need to create a `Program_indicator` for the `data` - don't have ID. Just pretty much that specific case hehe. Everything else all good we on same page!

**Comment by wlthomson** [comment]
> Ya sure - I'm just thinking about the value records as the example. `Value` comes in -> have ID for a `Data`, create placeholder. Need to create a `Program_indicator` for the `data` - don't have ID. Just pretty much that specific case hehe. Everything else all good we on same page!

Yeah, think we on the same page. `ProgramData` has its parent indicator ID, so I think that's all good, but I get the general idea 👍 .

**Comment by wlthomson** [comment]
Update: after discussing with bossman, updating `IndicatorData` to `IndicatorAttribute` (I think much clearer). Agree?

**Comment by joshxg** [comment]
Sure. Or Vector (i'm joking, attribute is good)

**Comment by wlthomson** [comment]
Fixed by #1846.
